src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/RazorProjectService.cs

```diff
@@ -139,41 +139,90 @@ await AddOrUpdateProjectCoreAsync(
             .ConfigureAwait(false);
     }
 
+    public async Task AddDocumentsToMiscProjectAsync(ImmutableArray<string> filePaths, CancellationToken cancellationToken)
+    {
+        await WaitForInitializationAsync().ConfigureAwait(false);
+
+        await _projectManager
+            .UpdateAsync(
+                (updater, cancellationToken) =>
+                {
+                    var projects = _projectManager.GetProjects();
+
+                    // For each file, check to see if it's already in a project.
+                    // If it is, we don't want to add it to the misc project.
+                    foreach (var filePath in filePaths)
+                    {
+                        foreach (var project in projects)
```

Comment:
This looks very similar to `TryFindContainingProject`

Reply:
It definitely is. However, calling `TryFindContainingProject` means creating a new array of `ProjectSnapshot` for every file. To avoid those fully avoidable allocations, I call `GetProjects()` once here.

Reply:
Ah I forgot `GetProjects()` was allocating. Thank you :)